src/Dialect/ONNX/ONNXOps/Canonicalize.cpp

```diff
@@ -1239,6 +1239,58 @@ class PowToMulRewritePattern : public OpRewritePattern<ONNXPowOp> {
   int64_t maxPower;
 };
 
+class PowConversionRewritePattern : public OpRewritePattern<ONNXPowOp> {
+public:
+  using OpRewritePattern<ONNXPowOp>::OpRewritePattern;
+
+  PowConversionRewritePattern(MLIRContext *context, int64_t maxPower)
+      : OpRewritePattern(context) {}
+
+  LogicalResult matchAndRewrite(
+      ONNXPowOp powOp, PatternRewriter &rewriter) const override {
+    Operation *op = powOp.getOperation();
+    Location loc = powOp.getLoc();
+
+    Value input = powOp.getX();
+    Value exp = powOp.getY();
+    if (!hasShapeAndRank(input) || !hasShapeAndRank(exp))
+      return failure(); // Cannot apply pattern until ranks are known.
+
+    // Characterize element types.
+    Type inputElementType =
+        mlir::cast<ShapedType>(input.getType()).getElementType();
+    Type expElementType =
+        mlir::cast<ShapedType>(exp.getType()).getElementType();
+    bool inputIsInt = isa<IntegerType>(inputElementType);
+    bool inputIsFloat = isa<FloatType>(inputElementType);
+    bool expIsInt = isa<IntegerType>(expElementType);
+    bool expIsFloat = isa<FloatType>(expElementType);
+    int64_t inputWidth = 0, expWidth = 0;
+    if (inputIsInt)
+      inputWidth = mlir::cast<IntegerType>(inputElementType).getWidth();
+    else if (inputIsFloat)
+      inputWidth = mlir::cast<FloatType>(inputElementType).getWidth();
+    if (expIsInt)
+      expWidth = mlir::cast<IntegerType>(expElementType).getWidth();
+    else if (expIsFloat)
+      expWidth = mlir::cast<FloatType>(expElementType).getWidth();
+
+    // Detect cases that MLIR supports without conversions => failure, aka no
+    // need to do anything here.
+    if ((inputWidth == expWidth) &&
+        ((inputIsInt && expIsInt) || (inputIsFloat && expIsFloat) ||
+            (inputIsFloat && expIsInt)))
+      return failure();
+
+    // Rewrite pow with a cast of the exp type to the input type.
+    MultiDialectBuilder<OnnxBuilder> create(rewriter, loc);
+    Value newExp = create.onnx.cast(exp, inputElementType);
```

Comment:
is it safe enough to always cast to the base operand's type? 

side question about `onnx.Pow` definition: what should be the result type width be? is it always the same as the base operand type width? if it is, i guess semantically it makes sense to always cast the exponent to base type.

Reply:
We currently don't check that the type are the ones supported by the standard. I assume it would complain at some later time if the input type was a string. I don't think we have done extensive tests for input types, so I am assuming that it would be good here too.

I believe matching the type of the base make sense as the type of the output is the same as the type of the base.